### ISSUE TITLE
Customize default database table name

### ISSUE THREAD

**Issue by xuanhnvt** [rhertogh/yii2-oauth2-server#5]
Hello,

I have scenario to build 2 separate oauth2 servers on the same database schema, so I have to customize default database table name for one. Is it possible to do that?

### What steps will reproduce the problem?
As I knew the default table name are generated from name of child class of base class rhertogh\Yii2Oauth2Server\models\base\Oauth2BaseActiveRecord in below part

```
abstract class Oauth2BaseActiveRecord extends ActiveRecord implements Oauth2ActiveRecordInterface
{
    use Oauth2ActiveRecordTrait;

    /**
     * @var string|null The name for the table
     */
    public static $tableName = null;

    /**
     * @inheritdoc
     */
    public static function tableName()
    {
        if (!empty(static::$tableName)) {
            return static::$tableName;
        }

        return '{{%' . Inflector::camel2id(StringHelper::basename(get_called_class()), '_') . '}}';
    }
...
}
```
I created all related models inheriting this base class and set static variable $tableName.

```
<?php

namespace app\modules\oauth2\models;

class Oauth2AccessToken extends \rhertogh\Yii2Oauth2Server\models\Oauth2AccessToken
{
     /**
     * @var string|null The name for the table
     */
    public static $tableName = 'contact_oauth2_access_token';

    /**
     * @return \rhertogh\Yii2Oauth2Server\interfaces\models\queries\Oauth2AccessTokenScopeQueryInterface     */
    public function getAccessTokenScopes()
    {
        return $this->hasMany(\app\modules\oauth2\models\Oauth2AccessTokenScope::className(), ['access_token_id' => 'id'])->inverseOf('accessToken');
    }

    /**
     * @return \rhertogh\Yii2Oauth2Server\interfaces\models\queries\Oauth2ClientQueryInterface     */
    public function getClient()
    {
        return $this->hasOne(\app\modules\oauth2\models\Oauth2Client::className(), ['id' => 'client_id'])->inverseOf('accessTokens');
    }

    /**
     * @return \rhertogh\Yii2Oauth2Server\interfaces\models\queries\Oauth2RefreshTokenQueryInterface     */
    public function getRefreshTokens()
    {
        return $this->hasMany(\app\modules\oauth2\models\Oauth2RefreshToken::className(), ['access_token_id' => 'id'])->inverseOf('accessToken');
    }
...
}
?>
```
Then I extended OAuth2Module, override default interface implementation with my custom models as below

```
namespace app\modules\oauth2;

use Yii;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2AccessTokenInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2AccessTokenScopeInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2AuthCodeInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2AuthCodeScopeInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2ClientInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2ClientScopeInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2RefreshTokenInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2ScopeInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2UserClientInterface;
use rhertogh\Yii2Oauth2Server\interfaces\models\Oauth2UserClientScopeInterface;
use app\modules\oauth2\models\Oauth2AccessToken;
use app\modules\oauth2\models\Oauth2AccessTokenScope;
use app\modules\oauth2\models\Oauth2AuthCode;
use app\modules\oauth2\models\Oauth2AuthCodeScope;
use app\modules\oauth2\models\Oauth2Client;
use app\modules\oauth2\models\Oauth2ClientScope;
use app\modules\oauth2\models\Oauth2RefreshToken;
use app\modules\oauth2\models\Oauth2Scope;
use app\modules\oauth2\models\Oauth2UserClient;
use app\modules\oauth2\models\Oauth2UserClientScope;

class Oauth2Module extends \rhertogh\Yii2Oauth2Server\Oauth2Module
{
    /**
     * Override default mapping for interfaces of models
     * @since 1.0.0
     */
    protected const OVERRIDE_DEFAULT_INTERFACE_IMPLEMENTATIONS = [
        # Models
        Oauth2AccessTokenInterface::class => Oauth2AccessToken::class,
        Oauth2AccessTokenScopeInterface::class => Oauth2AccessTokenScope::class,
        Oauth2AuthCodeInterface::class => Oauth2AuthCode::class,
        Oauth2AuthCodeScopeInterface::class => Oauth2AuthCodeScope::class,
        Oauth2ClientInterface::class => Oauth2Client::class,
        Oauth2ClientScopeInterface::class => Oauth2ClientScope::class,
        Oauth2RefreshTokenInterface::class => Oauth2RefreshToken::class,
        Oauth2ScopeInterface::class => Oauth2Scope::class,
        Oauth2UserClientInterface::class => Oauth2UserClient::class,
        Oauth2UserClientScopeInterface::class => Oauth2UserClientScope::class,
    ];

    /**
     * @inheritDoc
     * @throws InvalidConfigException
     */
    public function init()
    {
        parent::init();
        
        foreach (static::OVERRIDE_DEFAULT_INTERFACE_IMPLEMENTATIONS as $interface => $implementation) {
            Yii::$container->set($interface, $implementation);
        }
    }
}

```
Then specify my module class Oauth2Module in config.

### What is the expected result?
Generate migration, run command migrate, create set of tables with my custom name.
Server can authorize client without any error.

### What do you get instead?
Can create set of tables with my custom name but can not authorize client. It happened error after I did login

`{"error":"PHP Warning","error_description":"yii\\base\\ErrorException: get_class() expects parameter 1 to be object, null given in /.../vendor/rhertogh/yii2-oauth2-server/src/components/repositories/base/Oauth2BaseTokenRepository.php:82\nStack trace:\n#0 [internal function]: yii\\base\\ErrorHandler->handleError(2, 'get_class() exp...', '/Users/hungxuan...', 82, Array)\n#1 /.../vendor/rhertogh/yii2-oauth2-server/src/components/repositories/base/Oauth2BaseTokenRepository.php(82): get_class(NULL)\n#2 /.../vendor/rhertogh/yii2-oauth2-server/src/components/repositories/Oauth2AuthCodeRepository.php(37): rhertogh\\Yii2Oauth2Server\\components\\repositories\\base\\Oauth2BaseTokenRepository->persistToken(Object(app\\modules\\oauth2\\models\\Oauth2AuthCode))\n#3 /.../vendor/league/oauth2-server/src/Grant/AbstractGrant.php(514): rhertogh\\Yii2Oauth2Server\\components\\repositories\\Oauth2AuthCodeRepository->persistNewAuthCode(Object(app\\modules\\oauth2\\models\\Oauth2AuthCode))\n#4 /.../vendor/league/oauth2-server/src/Grant/AuthCodeGrant.php(353): League\\OAuth2\\Server\\Grant\\AbstractGrant->issueAuthCode(Object(DateInterval), Object(app\\modules\\oauth2\\models\\Oauth2Client), 801, 'https://oauth.p...', Array)\n#5 /.../vendor/league/oauth2-server/src/AuthorizationServer.php(181): League\\OAuth2\\Server\\Grant\\AuthCodeGrant->completeAuthorizationRequest(Object(League\\OAuth2\\Server\\RequestTypes\\AuthorizationRequest))\n#6 /.../vendor/rhertogh/yii2-oauth2-server/src/controllers/web/server/Oauth2AuthorizeAction.php(316): League\\OAuth2\\Server\\AuthorizationServer->completeAuthorizationRequest(Object(League\\OAuth2\\Server\\RequestTypes\\AuthorizationRequest), Object(GuzzleHttp\\Psr7\\Response))\n#7 [internal function]: rhertogh\\Yii2Oauth2Server\\controllers\\web\\server\\Oauth2AuthorizeAction->run('ITEkVow0DUl7J2y...')\n#8 /.../vendor/yiisoft/yii2/base/Action.php(93): call_user_func_array(Array, Array)\n#9 /.../vendor/yiisoft/yii2/base/Controller.php(178): yii\\base\\Action->runWithParams(Array)\n#10 /.../vendor/yiisoft/yii2/base/Module.php(552): yii\\base\\Controller->runAction('authorize', Array)\n#11 /.../vendor/yiisoft/yii2/web/Application.php(103): yii\\base\\Module->runAction('oauth2/server/a...', Array)\n#12 /.../vendor/yiisoft/yii2/base/Application.php(384): yii\\web\\Application->handleRequest(Object(yii\\web\\Request))\n#13 /.../web/index.php(12): yii\\base\\Application->run()\n#14 {main}"}`

I'm not sure this is correct solution. If I follow this way, is there any issue in the future or can suggest me easy solution to sort it out?

Thanks!

**Comment by rhertogh** [comment]
Hi @xuanhnvt,
The table names are indeed static, the function is inherited from `\yii\db\ActiveRecord::tableName()` so it probably won't be a easy task to have different database tables for the same class.
Are you sure you need 2 servers? What is the purpose of it?

**Comment by xuanhnvt** [comment]
Hi @rhertogh,

Thanks for your reply.

Currently, I have single database, it stores 2 types of login credential: internal user and customer. I need one oauth2 server for each:
- One server authorize internal user for internal apps.
- The other authorize customer for customer-specific apps only

If you have any ideas for my use case, kindly suggest. Thanks!

**Comment by rhertogh** [comment]
Hi @xuanhnvt 
I've just released version [1.0.0-alpha6](https://github.com/rhertogh/yii2-oauth2-server/blob/master/CHANGELOG.md#100-alpha6---2022-11-13)
The `Oauth2UserInterface` now defines a `isOauth2ClientAllowed()` by which the oauth server calls to determine if the User can access the client.
In this way you can implement the method on the User model and control which users can use specific clients.
I think this should solve your requirements without the need for 2 separate install by simply defining 2 oauth clients.

**Comment by xuanhnvt** [comment]
Hi @rhertogh

It's so great. That's what I need, thank you so much.
